### PR TITLE
fix: Incorporate `basePath` correctly in `useRouter` functions

### LINKED ISSUE #905: Router push loses locale with basePath
### Description

If a `basePath` is configured in the `nextConfig` it gets lost with `router.push`. With a `Link` it works fine.

### Mandatory reproduction URL

https://codesandbox.io/p/devbox/great-jerry-s6ylgc

### Reproduction description

Steps to reproduce:
1. Open reproduction
2. You should be now on` /base-path/de`
3. Click on Link
4. you are now on `/base-path/de/test-page`
5. go back
6. Click on Router Push
7. you are now on `/base-path/test-page`

### Expected behaviour

If using` router.push` in combination with `basePath` the locale should also be carried with.

### LINKED ISSUE #910: Navigation API router not working when Next.js is configured with a basepath
### Description

I have set up a project using Next.js 14 (App Router) and have been trying to integrate with **next-intl** by reading the docs. I am mocking up a simple application with a dashboard page (`/dashboard`) and a login page (`/login`) and I want to have internationalization handled.

There is a pseudo-request being performed (a timeout task that resolves a promise) on the client-side that returns a route to redirect to when the user navigates to a non-login page (mimicking a request response which determines that the current user does not have an active logged-in session yet). Since redirection needs to happen on the client side for the project I am working on, I need to use the navigation API rather than the middleware.

When not using a `basePath` in the Next.js config, the code seems to work as expected. However, when setting a basePath (e.g. `/admin`) and navigating say to the home page (`/admin`), the redirection that should take the user to the login page at `/admin/en/login`, fails. Instead, the user is redirected to `/admin/login`, which is not an existing page.

I tested on my larger project and `<Link>` components generated from the navigation API seem to work still. It seems that the router being returned by the `useRouter()` hook is the one having the issue.

**Pages**
1. Home (`./app/page.jsx` or `./app/[locale]/page.jsx`)
2. Dashboard (`./app/[locale]/dashboard/page.jsx`)
3. Login (`./app/[locale]/login/page.jsx`)

**Redirect Component**
See the _Promise_ defined inside the `useEffect()` hook call in `./components/AuthRouter/authRouter.js`.

### Mandatory reproduction URL

https://codesandbox.io/p/devbox/cranky-hamilton-8yz98j?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt946y7c00073b6i80upd3bp%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt946y7b00023b6irk8st3ib%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt946y7b00043b6ihwfz33xc%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt946y7c00063b6i3tsojyp8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt946y7b00023b6irk8st3ib%2522%253A%257B%2522id%2522%253A%2522clt946y7b00023b6irk8st3ib%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clt946y7c00063b6i3tsojyp8%2522%253A%257B%2522id%2522%253A%2522clt946y7c00063b6i3tsojyp8%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cltdcgp6f02ir3b6irsu12ft6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252Fadmin%252Flogin%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cltdcgp6f02ir3b6irsu12ft6%2522%257D%252C%2522clt946y7b00043b6ihwfz33xc%2522%253A%257B%2522id%2522%253A%2522clt946y7b00043b6ihwfz33xc%2522%252C%2522activeTabId%2522%253A%2522cltdh9gkn02en3b6iu50mnqmp%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522cltdh9gkn02en3b6iu50mnqmp%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

### Reproduction description

Steps to reproduce:
1. Navigate to code sandbox using reproduction link provided.
2. Start the Dev server if not started on its own.
3. Go to `/admin`. You'll notice an error show up.
4. Similarly go to `/admin/en` or `/admin/en/dashboard` and you'll see the same issue.
5. In contrast, navigate directly to `/admin/en/login` and you'll see the login page show up just fine.

### Expected behaviour

The router created by the navigation API should be able to correctly route from non-login pages to the login page when a `basePath` is set in the Next.js configuration file.

### PR DESCRIPTION
**Pull request by Oberwaditzer**
Fixes #905 
Fixes #910

### CODE REVIEW

**Review by amannn — COMMENTED** [review]
Hey, thanks for proposing this PR! Can you add a regression test that demonstrates the previously broken behavior?

**Comment by Oberwaditzer** [comment]
Hey @amannn,
I added some tests. If something is missing, please let me know :)

File: packages/next-intl/src/shared/utils.tsx [hunk 1]
```diff
@@ -90,7 +90,11 @@ export function prefixPathname(locale: string, pathname: string) {
 
 export function hasPathnamePrefixed(locale: string, pathname: string) {
   const prefix = `/${locale}`;
-  return pathname === prefix || pathname.startsWith(`${prefix}/`);
+  if (pathname === prefix) return true;
+  const hasPrefix = pathname.startsWith(prefix);
+  // check if the locale is the second segment of the URL to support basePath
+  if (!hasPrefix && pathname.split('/')?.[2] === locale) return true;
```
Comment (amannn) [inline packages/next-intl/src/shared/utils.tsx:96]: The base path can contain multiple segments (e.g. `/base/path`), we need to be a bit more flexible here.

I found a way to detect the base path like this:

https://github.com/amannn/next-intl/blob/16b19e861dcec88fc6e436cadf99543340819bcd/packages/next-intl/src/navigation/shared/syncLocaleCookie.tsx#L28

Maybe that can be used when router functions are called?

Comment (Oberwaditzer) [reply packages/next-intl/src/shared/utils.tsx]: you are right, haven't thought about that. I added a similar version to the baseRouter.

**Comment by amannn** [comment]
Thanks for adding the tests! I had another look and found a few things to clean up along the way: https://github.com/Oberwaditzer/next-intl/pull/1. Mind merging this into your branch?

**Comment by Oberwaditzer** [comment]
> Thanks for adding the tests! I had another look and found a few things to clean up along the way: [Oberwaditzer#1](https://github.com/Oberwaditzer/next-intl/pull/1). Mind merging this into your branch?

@amannn Done :)

**Review by amannn — APPROVED** [review]
Thank you so much @Oberwaditzer! 🙌

**Comment by Oberwaditzer** [comment]
@amannn The world is small.. 😅 But very cool to see that someone from little Vorarlberg has created such a library! :)

**Pull request merged**